File: cc/google/fhir/stu3/json_parser.cc
/*
 * Copyright 2018 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "google/fhir/stu3/json_format.h"

#include <ctype.h>
#include <iosfwd>
#include <memory>
#include <unordered_map>
#include <utility>

#include "google/protobuf/descriptor.h"
#include "google/protobuf/message.h"
#include "absl/memory/memory.h"
#include "absl/strings/str_cat.h"
#include "google/fhir/status/status.h"
#include "google/fhir/status/statusor.h"
#include "google/fhir/stu3/extensions.h"
#include "google/fhir/stu3/primitive_wrapper.h"
#include "google/fhir/stu3/util.h"
#include "proto/stu3/annotations.pb.h"
#include "proto/stu3/datatypes.pb.h"
#include "proto/stu3/google_extensions.pb.h"
#include "proto/stu3/resources.pb.h"
#include "include/json/json.h"
#include "tensorflow/core/lib/core/errors.h"
#include "re2/re2.h"

namespace google {
namespace fhir {
namespace stu3 {

namespace {

using ::google::fhir::Status;
using ::google::fhir::StatusOr;
using ::google::fhir::stu3::ClearTypedExtensions;
using ::google::fhir::stu3::IsChoiceType;
using ::google::fhir::stu3::IsPrimitive;
using ::google::fhir::stu3::IsReference;
using ::google::fhir::stu3::IsResource;
using ::google::fhir::stu3::ReferenceProtoToString;
using ::google::fhir::stu3::ReferenceStringToProto;
using ::google::fhir::stu3::proto::ContainedResource;
using ::google::fhir::stu3::proto::Extension;
using ::google::protobuf::Descriptor;
using ::google::protobuf::FieldDescriptor;
using ::google::protobuf::Message;
using ::google::protobuf::Reflection;
using ::tensorflow::errors::InvalidArgument;

// Given a proto descriptor, constructs a map from FHIR JSON field name to
// FieldDescriptor.
// Since FHIR represents extensions to primitives as separate JSON fields,
// prepended by underscore, we add that as a separate mapping to the primitive
// field.
// TODO: memoize
std::unordered_map<string, const FieldDescriptor*> GetFieldMap(
    const Descriptor* descriptor) {
  std::unordered_map<string, const FieldDescriptor*> field_map;

  for (int i = 0; i < descriptor->field_count(); i++) {
    const FieldDescriptor* field = descriptor->field(i);
    if (IsChoiceType(field)) {
      std::unordered_map<string, const FieldDescriptor*> inner_map =
          GetFieldMap(field->message_type());
      for (auto iter = inner_map.begin(); iter != inner_map.end(); iter++) {
        string inner_field_name = iter->first;
        inner_field_name[0] = std::toupper(inner_field_name[0]);
        field_map[absl::StrCat(field->json_name(), inner_field_name)] = field;
      }
    } else {
      field_map[field->json_name()] = field;
      if (field->type() == FieldDescriptor::TYPE_MESSAGE &&
          IsPrimitive(field->message_type())) {
        // Fhir JSON represents extensions to primitive fields as separate
        // standalone JSON objects, keyed by the "_" + field name.
        field_map["_" + field->json_name()] = field;
      }
    }
  }
  return field_map;
}

// Builds a map from ContainedResource field name to FieldDescriptor for that
// field.
std::unordered_map<string, const FieldDescriptor*>* BuildResourceTypeMap() {
  std::unordered_map<string, const FieldDescriptor*>* map =
      new std::unordered_map<string, const FieldDescriptor*>;
  for (int i = 0; i < ContainedResource::descriptor()->field_count(); i++) {
    const FieldDescriptor* field = ContainedResource::descriptor()->field(i);
    (*map)[field->message_type()->name()] = field;
  }
  return map;
}

static const std::unordered_map<string, const FieldDescriptor*>*
    kResourceTypeMap = BuildResourceTypeMap();

class Parser {
 public:
  explicit Parser(absl::TimeZone default_timezone)
      : default_timezone_(default_timezone) {}

  Status MergeMessage(const Json::Value& value, Message* target) {
    const Descriptor* target_descriptor = target->GetDescriptor();
    if (target_descriptor->full_name() ==
        ContainedResource::descriptor()->full_name()) {
      // We handle contained resources in a special way, because despite
      // internally being a Oneof, it is not acually a choice-type in FHIR. The
      // JSON field name is just "resource", which doesn't give us any clues
      // about which field in the Oneof to set.  Instead, we need to inspect
      // the JSON input to determine its type.  Then, merge into that specific
      // field in the resource Oneof.
      string resource_type =
          value.get("resourceType", Json::Value::null).asString();
      auto resource_field_iter = kResourceTypeMap->find(resource_type);
      if (resource_field_iter == kResourceTypeMap->end()) {
        return InvalidArgument("Unsupported resource type: ", resource_type);
      }
      return MergeMessage(value, target->GetReflection()->MutableMessage(
                                     target, resource_field_iter->second));
    }

    const std::unordered_map<string, const FieldDescriptor*> field_map =
        GetFieldMap(target_descriptor);

    for (auto sub_value_iter = value.begin(); sub_value_iter != value.end();
         ++sub_value_iter) {
      const auto& field_entry = field_map.find(sub_value_iter.key().asString());
      if (field_entry != field_map.end()) {
        if (field_entry->second->options().GetExtension(
                ::google::fhir::stu3::proto::is_choice_type)) {
          FHIR_RETURN_IF_ERROR(MergeChoiceField(*sub_value_iter,
                                                field_entry->second,
                                                field_entry->first, target));
        } else {
          FHIR_RETURN_IF_ERROR(
              MergeField(*sub_value_iter, field_entry->second, target));
        }
      } else if (sub_value_iter.key().asString() == "resourceType") {
        string resource_type = sub_value_iter->asString();
        if (!IsResource(target_descriptor) ||
            target_descriptor->name() != resource_type) {
          return InvalidArgument("Error merging json resource of type ",
                                 resource_type, " into message of type",
                                 target_descriptor->name());
        }
      } else {
        return InvalidArgument("Unable to merge field ",
                               absl::StrCat(sub_value_iter.key().asString()),
                               " into resource of type ",
                               target_descriptor->full_name());
      }
    }
    return Status::OK();
  }

  Status MergeChoiceField(const Json::Value& json,
                          const FieldDescriptor* choice_field,
                          const string& field_name, Message* parent) {
    const Descriptor* choice_type_descriptor = choice_field->message_type();
    string value_suffix = field_name.substr(choice_field->json_name().length());
    value_suffix[0] = std::tolower(value_suffix[0]);
    const auto choice_type_field_map = GetFieldMap(choice_type_descriptor);
    auto value_field_iter = choice_type_field_map.find(value_suffix);
    if (value_field_iter == choice_type_field_map.end()) {
      return InvalidArgument("Can't find ", value_suffix, " on ",
                             choice_field->full_name());
    }
    Message* choice_msg =
        parent->GetReflection()->MutableMessage(parent, choice_field);
    return MergeField(json, value_field_iter->second, choice_msg);
  }

  // Given a JSON value, field, and parent message, merges the FHIR JSON into
  // the given field on the parent.
  // Note that we cannot just pass the field message, as this behaves
  // differently if the field has been previously set or not.
  Status MergeField(const Json::Value& json, const FieldDescriptor* field,
                    Message* parent) {
    const Reflection* parent_reflection = parent->GetReflection();
    // If the field is non-primitive make sure it hasn't been set yet.
    // Note that we allow primitive types to be set already, because FHIR
    // represents extensions to primitives as separate, subsequent JSON
    // elements, with the field prepended by an underscore.  In #GetFieldMap
    // above, these were mapped to the same fields.
    if (!IsPrimitive(field->message_type()) &&
        !(field->is_repeated() &&
          parent_reflection->FieldSize(*parent, field) == 0) &&
        !(!field->is_repeated() &&
          !parent_reflection->HasField(*parent, field))) {
      return InvalidArgument("Target field already set: ", field->full_name(),
                             "\n", parent->DebugString(), "\n",
                             field->full_name(), "\n",
                             absl::StrCat(json.toStyledString()), "\n done");
    }

    if (field->containing_oneof()) {
      const ::google::protobuf::FieldDescriptor* oneof_field =
          parent_reflection->GetOneofFieldDescriptor(*parent,
                                                     field->containing_oneof());
      if (oneof_field) {
        return InvalidArgument(
            "Cannot set field ", field->full_name(), " because another field ",
            oneof_field->full_name(), " of the same oneof is already set.");
      }
    }

    if (field->is_repeated()) {
      if (!json.isArray()) {
        return InvalidArgument(
            "Attempted to set repeated field ", field->full_name(),
            " using non-array JSON: ", absl::StrCat(json.toStyledString()));
      }
      size_t existing_field_size = parent_reflection->FieldSize(*parent, field);
      if (existing_field_size != 0 && existing_field_size != json.size()) {
        return InvalidArgument(
            "Repeated primitive list length does not match extension list ",
            "for field: ", field->full_name());
      }
      for (Json::ArrayIndex i = 0; i < json.size(); i++) {
        FHIR_ASSIGN_OR_RETURN(
            std::unique_ptr<Message> parsed_value,
            ParseFieldValue(field, json.get(i, Json::Value::null), parent));
        if (existing_field_size > 0) {
          Message* field_value =
              parent_reflection->MutableRepeatedMessage(parent, field, i);
          field_value->MergeFrom(*parsed_value);
          // This is the second time we've visited this field - once for
          // extensions, and once for value.  So, make sure to clear the
          // PrimitiveHasNoValue extension.
          FHIR_RETURN_IF_ERROR(ClearPrimitiveHasNoValue(field_value));
        } else {
          parent_reflection->AddAllocatedMessage(parent, field,
                                                 parsed_value.release());
        }
      }
    } else {
      FHIR_ASSIGN_OR_RETURN(std::unique_ptr<Message> parsed_value,
                            ParseFieldValue(field, json, parent));
      if (parent_reflection->HasField(*parent, field)) {
        Message* field_value = parent_reflection->MutableMessage(parent, field);
        field_value->MergeFrom(*parsed_value);
        // This is the second time we've visited this field - once for
        // extensions, and once for value.  So, make sure to clear the
        // PrimitiveHasNoValue extension.
        FHIR_RETURN_IF_ERROR(ClearPrimitiveHasNoValue(field_value));
      } else {
        parent_reflection->SetAllocatedMessage(parent, parsed_value.release(),
                                               field);
      }
    }
    return Status::OK();
  }

  Status ClearPrimitiveHasNoValue(Message* message) {
    return ClearTypedExtensions(
        ::google::fhir::stu3::google::PrimitiveHasNoValue::descriptor(),
        message);
  }

  StatusOr<std::unique_ptr<Message>> ParseFieldValue(
      const FieldDescriptor* field, const Json::Value& json, Message* parent) {
    if (field->type() != FieldDescriptor::Type::TYPE_MESSAGE) {
      return InvalidArgument("Error in FHIR proto definition: Field ",
                             field->full_name(), " is not a message.");
    }
    std::unique_ptr<Message> target =
        absl::WrapUnique(parent->GetReflection()
                             ->GetMessageFactory()
                             ->GetPrototype(field->message_type())
                             ->New());
    FHIR_RETURN_IF_ERROR(MergeValue(json, target.get()));
    return std::move(target);
  }

  Status MergeValue(const Json::Value& json, Message* target) {
    if (IsPrimitive(target->GetDescriptor())) {
      if (json.isObject()) {
        // This is a primitive type extension.
        // Merge the extension fields into into the empty target proto,
        // and tag it as having no value.
        FHIR_RETURN_IF_ERROR(MergeMessage(json, target));
        target->GetReflection()
            ->AddMessage(target,
                         target->GetDescriptor()->FindFieldByName("extension"))
            ->CopyFrom(*GetPrimitiveHasNoValueExtension());
        return Status::OK();
      } else {
        FHIR_RETURN_IF_ERROR(ParseInto(json, default_timezone_, target));
        return Status::OK();
      }
    } else if (IsReference(target->GetDescriptor())) {
      ::google::fhir::stu3::proto::Reference reference;
      FHIR_RETURN_IF_ERROR(MergeMessage(json, &reference));

      // Proto references support more structure than json references.
      // Serialize and deserialize to get the most structured version
      // possible.
      StatusOr<string> reference_string = ReferenceProtoToString(reference);
      if (!reference_string.ok()) {
        // This code path is taken for References which don't have "reference"
        // set, such as references with only a "display" field.
        target->MergeFrom(reference);
        return Status::OK();
      } else {
        string reference_value = reference_string.ValueOrDie();
        FHIR_ASSIGN_OR_RETURN(
            ::google::fhir::stu3::proto::Reference parsed_reference,
            ReferenceStringToProto(reference_value));
        if (reference.has_display()) {
          parsed_reference.mutable_display()->set_value(
              reference.display().value());
        }
        if (reference.has_id()) {
          parsed_reference.mutable_id()->set_value(reference.id().value());
        }
        for (const Extension& extension : reference.extension()) {
          *(parsed_reference.add_extension()) = extension;
        }
        target->MergeFrom(parsed_reference);
        return Status::OK();
      }
    }
    // Must be another FHIR element.
    if (!json.isObject()) {
      return InvalidArgument("Expected JsonObject for field of type ",
                             target->GetDescriptor()->full_name());
    }
    FHIR_RETURN_IF_ERROR(MergeMessage(json, target));
    return Status::OK();
  }

 private:
  absl::TimeZone default_timezone_;
};

StatusOr<Json::Value> ParseJsonValue(const string& raw_json) {
  Json::Reader reader;
  Json::Value value;
  if (!reader.parse(raw_json, value)) {
    return InvalidArgument("Failed parsing raw json: ", raw_json);
  }
  return value;
}

}  // namespace

Status MergeJsonFhirStringIntoProto(const string& raw_json, Message* target,
                                    const absl::TimeZone default_timezone) {
  string mutable_raw_json = raw_json;
  // FHIR JSON format stores decimals as unquoted rational numbers.  This is
  // problematic, because their representation could change when they are
  // parsed into C++ doubles.  To avoid this, add quotes around any decimal
  // fields to ensure that they are parsed as strings.  Note that any field
  // that is already in quotes will not match this regex, and thus be ignored.
  //
  // This regex is three capture groups:
  // 1: a un-escaped double-quote followed by a colon and arbitrary whitespace,
  //    to ensure this is a field value (and not inside a string).
  // 2: a decimal
  // 3: any field-ending token.
  static const LazyRE2 kDecimalKeyValuePattern{
      "(?m)([^\\\\]\":\\s*)(-?\\d*\\.\\d*?)([\\s,\\}\\]$])"};
  RE2::GlobalReplace(&mutable_raw_json, *kDecimalKeyValuePattern,
                     "\\1\"\\2\"\\3");

  Json::Value value;

  // TODO: Decide if we want to support value-only JSON
  if (target->GetDescriptor()->full_name() ==
          proto::Decimal::descriptor()->full_name() &&
      raw_json != "null") {
    // Similar to above, if this is a standalone decimal, parse it as a string
    // to avoid changing reprentation due to precision.
    FHIR_ASSIGN_OR_RETURN(
        value, ParseJsonValue(absl::StrCat("\"", mutable_raw_json, "\"")));
  } else {
    FHIR_ASSIGN_OR_RETURN(value, ParseJsonValue(mutable_raw_json));
  }

  Parser parser{default_timezone};
  return parser.MergeValue(value, target);
}

}  // namespace stu3
}  // namespace fhir
}  // namespace google

File: py/google/fhir/json_format/_json_parser.py
#
# Copyright 2020 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""Functionality for parsing FHIR JSON into protobuf format."""

import threading
from typing import Any, Dict

from google.protobuf import any_pb2
from google.protobuf import descriptor
from google.protobuf import message
from google.fhir import extensions
from google.fhir import primitive_handler
from google.fhir import references
from google.fhir.utils import annotation_utils
from google.fhir.utils import proto_utils

_field_map_memos = {}
_field_map_memos_cv = threading.Condition(threading.Lock())


def _get_nested_choice_field_name(field: descriptor.FieldDescriptor,
                                  child_field_name: str) -> str:
  """Outputs the proper representation of field_name for a Choice type."""
  if child_field_name.startswith('_'):
    # For primitive extensions, prepend the leading underscore, e.g.:
    # value + _boolean = _valueBoolean
    return ('_' + field.json_name + child_field_name[1].upper() +
            child_field_name[1:])
  else:
    # Otherwise, just append together the JSON name, e.g.:
    # value + boolean = valueBoolean
    return (field.json_name + child_field_name[0].upper() +
            child_field_name[1:])


def _get_choice_field_name(field: descriptor.FieldDescriptor,
                           nested_field_name: str) -> str:
  """Returns the Choice type field name from a nested field name."""
  base_index = len(field.json_name)
  if nested_field_name.startswith('_'):
    # For primitive extensions, prepent the leading underscore, e.g.:
    # _valueBoolean = boolean
    return ('_' + nested_field_name[base_index + 1].lower() +
            nested_field_name[base_index + 2:])
  else:
    # Otherwise, just extract the field name, e.g.:
    # valueBoolean = boolean
    return (nested_field_name[base_index].lower() +
            nested_field_name[base_index + 1:])


def _get_field_map(
    desc: descriptor.Descriptor) -> Dict[str, descriptor.FieldDescriptor]:
  """Returns a mapping between field name and FieldDescriptor.

  Note that for FHIR "Choice" types, field names of primitive extensions are
  prepended with a leading underscore, e.g.: value + _boolean = _valueBoolean.

  Args:
    desc: The Descriptor of the Message whose mapping to return.

  Returns:
    A mapping between the field name and its corresponding FieldDescriptor.
  """
  # Early exit if we've already constructed the field mapping
  with _field_map_memos_cv:
    if desc in _field_map_memos:
      return _field_map_memos[desc]

  # Build field mapping
  field_map = {}
  for field in desc.fields:
    if (field.type == descriptor.FieldDescriptor.TYPE_MESSAGE and
        annotation_utils.is_choice_type_field(field)):
      inner_map = _get_field_map(field.message_type)
      for (child_field_name, _) in inner_map.items():
        choice_field_name = _get_nested_choice_field_name(
            field, child_field_name)
        field_map[choice_field_name] = field
    else:
      field_map[field.json_name] = field

      # FHIR JSON represents extensions to primitive fields as separate
      # standalone JSON objects, keyed by '_' + field_name
      if (field.type == descriptor.FieldDescriptor.TYPE_MESSAGE and
          annotation_utils.is_primitive_type(field.message_type)):
        field_map['_' + field.json_name] = field

  # Cache result
  with _field_map_memos_cv:
    _field_map_memos[desc] = field_map
    return _field_map_memos[desc]


class JsonParser:
  """A JSON to FHIR protobuf parser."""

  @classmethod
  def json_parser_with_default_timezone(
      cls, primitive_handler_: primitive_handler.PrimitiveHandler,
      default_timezone: str):
    """Returns a new parser initialized with the default_timezone."""
    return cls(primitive_handler_, default_timezone)

  def __init__(self, primitive_handler_: primitive_handler.PrimitiveHandler,
               default_timezone: str):
    """Initializes an instance of the FHIR JSON parser.

    Note that this is for *internal-use* only. External clients should leverage
    one of the available class constructors, such as:
    `JsonParser.json_parser_with_default_timezone(...)`.

    Args:
      primitive_handler_: Responsible for returning PrimitiveWrappers.
      default_timezone: The string representation of the timezone to default-to
        when parsing time-like values.
    """
    self.primitive_handler = primitive_handler_
    self.default_timezone = default_timezone

    # Mapping from field name to field on the appropriate ContainedResource
    self._resource_type_mapping = {
        field.message_type.name: field
        for field in primitive_handler_.contained_resource_cls.DESCRIPTOR.fields
    }

  def _parse_field_value(self, field: descriptor.FieldDescriptor,
                         json_value: Any) -> message.Message:
    """Returns a new Message described by the FieldDescriptor and json_value.

    Args:
      field: The FieldDescriptor of the Message instance to create.
      json_value: The JSON value representation to merge into the newly created
        Message.

    Returns:
      A new Message as described by the provided FieldDescriptor merged with the
      contents of json_value.
    """
    if field.type != descriptor.FieldDescriptor.TYPE_MESSAGE:
      raise ValueError('Error in FHIR proto definition, field: '
                       f'{field.full_name} is not a message.')
    if field.message_type.full_name == any_pb2.Any.DESCRIPTOR.full_name:
      contained = self.primitive_handler.new_contained_resource()
      self._merge_contained_resource(json_value, contained)

      any_message = any_pb2.Any()
      any_message.Pack(contained)
      return any_message
    else:
      target = proto_utils.create_message_from_descriptor(field.message_type)
      self.merge_value(json_value, target)
      return target

  def _merge_choice_field(self, json_value: Any,
                          choice_field: descriptor.FieldDescriptor,
                          field_name: str, parent: message.Message):
    """Creates a Message based on the choice_field Descriptor and json_value.

    The resulting message is merged into parent.

    Args:
      json_value: The JSON value to merge into a message of the type described
        by choice_field.
      choice_field: The field descriptor of the FHIR choice type on parent.
      field_name: The nested field name of the choice type, e.g.: _valueBoolean.
      parent: The parent Message to merge into.
    """
    choice_field_name = _get_choice_field_name(choice_field, field_name)
    choice_field_map = _get_field_map(choice_field.message_type)
    choice_value_field = choice_field_map.get(choice_field_name)
    if choice_value_field is None:
      raise ValueError(f'Cannot find {choice_field_name!r} on '
                       f'{choice_field.full_name}')

    choice_message = proto_utils.set_in_parent_or_add(parent, choice_field)
    self._merge_field(json_value, choice_value_field, choice_message)

  def _merge_field(self, json_value: Any, field: descriptor.FieldDescriptor,
                   parent: message.Message):
    """Merges the json_value into the provided field of the parent Message.

    Args:
      json_value: The JSON value to set.
      field: The FieldDescriptor of the field to set in parent.
      parent: The parent Message to set the value on.

    Raises:
      ValueError: In the event that a non-primitive field has already been set.
      ValueError: In the event that a oneof field has already been set.
    """
    # If the field is non-primitive, ensure that it hasn't been set yet. Note
    # that we allow primitive types to be set already, because FHIR represents
    # extensions to primitives as separate, subsequent JSON elements, with the
    # field prepended by an underscore
    if (not annotation_utils.is_primitive_type(field.message_type) and
        proto_utils.field_is_set(parent, field)):
      raise ValueError('Target field {} is already set.'.format(
          field.full_name))

    # Consider it an error if a oneof field is already set.
    #
    # Exception: When a primitive in a choice type has a value and an extension,
    # it will get set twice, once by the value (e.g. valueString) and once by an
    # extension (e.g., _valueString)
    if field.containing_oneof is not None:
      oneof_field = parent.DESCRIPTOR.oneofs_by_name[
          field.containing_oneof.name]
      if (annotation_utils.is_primitive_type(field.message_type) and
          oneof_field.full_name == field.full_name):
        raise ValueError(
            'Cannot set field {} since oneof field {} is already set.'.format(
                field.full_name, oneof_field.full_name))

    # Ensure that repeated fields get proper list assignment
    existing_field_size = proto_utils.field_content_length(parent, field)
    if proto_utils.field_is_repeated(field):
      if not isinstance(json_value, list):
        raise ValueError(
            f'Attempted to merge a repeated field, {field.name}, a json_value '
            f'with type {type(json_value)} instead of a list.')

      if existing_field_size != 0 and existing_field_size != len(json_value):
        raise ValueError(
            'Repeated primitive list length does not match extension list for '
            'field: {field.full_name!r}.')

    # Set the JSON values, taking care to clear the PRIMITIVE_HAS_NO_VALUE_URL
    # if we've already visited the field before.
    json_value = (
        json_value if proto_utils.field_is_repeated(field) else [json_value])
    for (i, value) in enumerate(json_value):
      parsed_value = self._parse_field_value(field, value)
      if existing_field_size > 0:
        field_value = proto_utils.get_value_at_field_index(parent, field, i)
        field_value.MergeFrom(parsed_value)
        extensions.clear_fhir_extensions_with_url(
            field_value, extensions.PRIMITIVE_HAS_NO_VALUE_URL)
      else:
        field_value = proto_utils.set_in_parent_or_add(parent, field)
        field_value.MergeFrom(parsed_value)

  def _merge_contained_resource(self, json_value: Dict[str, Any],
                                target: message.Message):
    """Merges json_value into a contained resource field within target."""
    # We handle contained resources in a special way, because despite internally
    # being a oneof, it is not actually a chosen-type in FHIR. The JSON field
    # name is just 'resource', which doesn't give any clues about which field in
    # the oneof to set. Instead, we need to inspect the JSON input to determine
    # its type. Then, merge into that specific fields in the resource oneof.
    resource_type = json_value.get('resourceType')
    contained_field = self._resource_type_mapping.get(resource_type)
    if contained_field is None:
      raise ValueError('Unable to retrieve ContainedResource field for '
                       f'{resource_type}.')
    contained_message = proto_utils.set_in_parent_or_add(
        target, contained_field)
    self._merge_message(json_value, contained_message)

  def _merge_message(self, json_value: Dict[str, Any], target: message.Message):
    """Merges the provided json object into the target Message."""
    target_descriptor = target.DESCRIPTOR
    if target_descriptor.name == 'ContainedResource':
      self._merge_contained_resource(json_value, target)
      return

    # For each field/value present in the JSON value, merge into target...
    field_map = _get_field_map(target_descriptor)
    for (field_name, value) in json_value.items():
      field = field_map.get(field_name)
      if field is not None:
        if annotation_utils.is_choice_type_field(field):
          self._merge_choice_field(value, field, field_name, target)
        else:
          self._merge_field(value, field, target)
      elif field_name == 'resourceType':
        if (not annotation_utils.is_resource(target_descriptor) or
            target_descriptor.name != value):
          raise ValueError(f'Error merging JSON resource into message of type '
                           f'{target_descriptor.name}.')
      else:
        raise ValueError(
            f'Unable to merge {field_name!r} into {target_descriptor.name}.')

  def merge_value(self, json_value: Any, target: message.Message):
    """Merges the provided json_value into the target Message.

    Args:
      json_value: A Python-native representation of JSON data.
      target: The target Message to merge the JSON data into.
    """
    target_descriptor = target.DESCRIPTOR
    if annotation_utils.is_primitive_type(target_descriptor):
      if isinstance(json_value, dict):
        # This is a primitive type extension. Merge the extension fields into
        # the empty target proto, and tag it as having no value.
        self._merge_message(json_value, target)

        extension_field = target_descriptor.fields_by_name.get('extension')
        if extension_field is None:
          raise ValueError("Invalid primitive. No 'extension' field exists on "
                           f"{target_descriptor.full_name}.")
        primitive_has_no_value = extensions.create_primitive_has_no_value(
            extension_field.message_type)
        proto_utils.append_value_at_field(target, extension_field,
                                          primitive_has_no_value)
      else:
        wrapper = self.primitive_handler.primitive_wrapper_from_json_value(
            json_value, type(target), default_timezone=self.default_timezone)
        wrapper.merge_into(target)
    elif annotation_utils.is_reference(target_descriptor):
      self._merge_message(json_value, target)
      references.split_if_relative_reference(target)
    else:
      if isinstance(json_value, dict):
        # The provided value must be another FHIR element
        self._merge_message(json_value, target)
      elif isinstance(json_value, (tuple, list)) and len(json_value) == 1:
        # Check if the target field is non-repeated, and we're trying to
        # populate it with a single-element array. This is considered valid, and
        # occurs when a profiled resource reduces the size of a repeated FHIR
        # field to a maximum of 1.
        self._merge_message(json_value[0], target)
      else:
        raise ValueError('Expected a JSON object for field of type: {}.'.format(
            target_descriptor.full_name))
